consensus/polybft/checkpoint_manager.go

```diff
@@ -246,8 +279,70 @@ func (c *checkpointManager) abiEncodeCheckpointBlock(blockNumber uint64, blockHa
 	return submitCheckpointMethod.Encode(params)
 }
 
-// isCheckpointBlock returns true for blocks in the middle of the epoch
-// which are offseted by predefined count of blocks
-func (c *checkpointManager) isCheckpointBlock(blockNumber uint64) bool {
-	return blockNumber == c.latestCheckpointID+c.checkpointsOffset
+// IsCheckpointBlock returns true for blocks in the middle of the epoch
+// which are offset by predefined count of blocks
+// or if given block is an epoch ending block
+func (c *checkpointManager) IsCheckpointBlock(blockNumber uint64, isEpochEndingBlock bool) bool {
+	return isEpochEndingBlock || blockNumber == c.latestCheckpointID+c.checkpointsOffset
+}
+
+// PostBlock is called on every insert of finalized block (either from consensus or syncer)
+// It will read any exit event that happened in block and insert it to state boltDb
+func (c *checkpointManager) PostBlock(req *PostBlockRequest) error {
+	epoch := req.Epoch
+	if req.IsEpochEndingBlock {
+		// exit events that happened in epoch ending blocks,
+		// should be added to the tree of the next epoch
+		epoch++
+	}
+
+	// commit exit events only when we finalize a block
+	events, err := getExitEventsFromReceipts(epoch, req.FullBlock.Block.Number(), req.FullBlock.Receipts)
+	if err != nil {
+		return err
+	}
+
+	if len(events) > 0 {
+		if err := c.state.insertExitEvents(events); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getExitEventsFromReceipts parses logs from receipts to find exit events
+func getExitEventsFromReceipts(epoch, block uint64, receipts []*types.Receipt) ([]*ExitEvent, error) {
+	events := make([]*ExitEvent, 0)
+
+	for i := 0; i < len(receipts); i++ {
+		if len(receipts[i].Logs) == 0 {
```

Comment:
You don need this check, range will work even with `nil` values

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/1113/commits/be68a84e56c50260e2d3fcf66bcc559212cc3375